### PR TITLE
✨ Add: disabling of csv button while data is incomplete (PCD-113)

### PR DESCRIPTION
**Pull request by sebasfavaron**
Miscellaneous changes:
+ eliminated react warning (unique key on children)
+ search now starts with 3 or more chars (previously 2 or more)
+ ens names error (mainly caused by emojis) now don't cause all other ens names to fail

## Summary

Csv download button was previously always available, leading users to download a csv with incomplete data. It now waits for all tokens and the ens names to be loaded.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)

## Screenshots

Screenshot from 2021-10-22 10-58-34

## Testing

Check if the button is correctly disabled during loading. Unfortunately we don't have a progress bar or something similar, so its kind of a guess, but you can download as soon as the button is enabled and much later to check for changes

## Ticket

[PCD-113](https://poap-devs.atlassian.net/jira/software/projects/PCD/boards/9/roadmap?assignee=5c4bc8d4043b4f5d172a6fdd&selectedIssue=PCD-113)
This PR applies to the ticket partially. The 'wrong linking to ens addresses' part is already fixed elsewhere

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I didn't commit unnecessary logs

### CODE REVIEW

File: src/components/eventCard.js [hunk 1]
```diff
@@ -69,8 +69,8 @@ function Content({type, width, size, event, power}) {
         }
     }, [event, transferCount])
 
-    const nl2br = (text) => (text.split('\n').map((item, key) => {
-        return <>{item}<br/></>
+    const nl2br = (text) => (text.split('\n').map(item => {
+        return <React.Fragment key={item.toString()}>{item}<br/></React.Fragment>
```
Comment (pablojp) [inline src/components/eventCard.js:73]: here we can use tags `<>`  and `</>` to replace React.Fragment and simplify the code

Comment (sebasfavaron) [reply src/components/eventCard.js:73]: to be able to add key prop

Comment (sebasfavaron) [reply src/components/eventCard.js:73]: to be able to add key prop

File: src/pages/event.js [hunk 2]
```diff
@@ -127,6 +133,22 @@ export function Event() {
...
+      // Finished last step, all token data is available (except for ens data)
+      setCanDownloadCsv(CanDownloadCsvStatus.ReadyWithoutEns)
+      let ownerIds = tokens.map(t => t.owner.id)
+      getEnsData(ownerIds).then(allnames => {
+        if(allnames.length > 0){
+          setEnsNames(allnames)
+        }
+        setCanDownloadCsv(CanDownloadCsvStatus.Ready)
+      }).catch((e) => {
+        console.error('Error getting ens data. ', e)
```
[the remark this answers is not in the document]
Comment (pablojp) [reply src/pages/event.js:147]: we need show a message to user?

File: src/pages/event.js [hunk 3]
```diff
@@ -231,25 +265,25 @@ export function Event() {
...
+              className={'btn'}
+              style={csvButtonStyle}
+              data={csv_data}
+            >
+              <span style={{margin: 0}}>{`Download CSV${canDownloadCsv === CanDownloadCsvStatus.ReadyWithoutEns ? ' (without ens)' : ''}`}</span>
+              <ReactTooltip effect={'solid'} />
+            </CSVLink> :
+            <button
+                className={'btn button-disabled'}
+                style={csvButtonStyle}
```
Comment (pablojp) [inline src/pages/event.js:282]: Why are we mixing style components with css classes?

File: src/components/eventCard.js (hunk 1 above)
Comment (pablojp) [reply src/components/eventCard.js:73]: ok, this must be a snippet of something in particular? Could it be handled with a <p> tag to make it semantically more correct in html code?

**Review by pablojp — CHANGES_REQUESTED** [review]

File: src/pages/event.js (hunk 3 above)
Comment (sebasfavaron) [reply src/pages/event.js:282]: the classnames are the way of stylizing buttons and the style is specific to this button

File: src/pages/gallery.js [hunk 4]
```diff
@@ -50,7 +49,7 @@ export default function Gallery() {
     debounceHandleSearch(value, items)
   };
   const handleNewSearchValue = (value, items) => {
-    if (value && value.length > 1) {
+    if (value && value.length > 2) {
```
[the remark this answers is not in the document]
Comment (sebasfavaron) [reply src/pages/gallery.js:52]: this was a bussiness request. "Only search with 3 or more chars"

File: package.json [hunk 5]
```diff
@@ -41,7 +42,8 @@
     "redux-persist": "^6.0.0",
     "redux-thunk": "^2.3.0",
     "request": "^2.88.2",
-    "serverless-http": "^2.7.0"
+    "serverless-http": "^2.7.0",
+    "styled-components": "^5.3.3"
```
Comment (pablojp) [inline package.json:46]: Why do you add fancy components? Can you use CSS classes defined in files? It is to follow the standard in the project and not add more handles of different styles.

File: src/pages/event.js [hunk 6]
```diff
@@ -127,6 +143,24 @@ export function Event() {
     }
   }, [ensNames]) /* eslint-disable-line react-hooks/exhaustive-deps */
 
+  useEffect(() => {
+    if (loadingEvent === 'succeeded' && csvDownloadIsOnLastStep()) {
```
Comment (pablojp) [inline src/pages/event.js:147]: We need this to be a constant `'succeeded'`. Move all values of loadingEvent to constants

File: src/pages/event.js [hunk 7]
```diff
@@ -127,6 +143,24 @@ export function Event() {
...
+      setCanDownloadCsv(CSV_STATUS.ReadyWithoutEns)
```
Comment (pablojp) [inline src/pages/event.js:148]: Can you move this logic from line 148 to 160, to a function called validationDownloadCSV or some related name?

File: src/pages/event.js [hunk 8]
```diff
@@ -190,77 +218,71 @@ export function Event() {
...
-            <div className='prev-next-buttons' style={{display: 'flex', justifyContent: 'space-between', marginBottom: 38,}}>
-              <a href={parseInt(eventId)-1} ><FontAwesomeIcon icon={faAngleLeft}/>{'  Prev'}</a>
-              <h4 style={{marginBottom: '0'}}><div className='event-title'>EVENT ID</div><div className='event-id'>#{eventId}</div> </h4>
-              <a href={parseInt(eventId)+1} >{'Next  '}<FontAwesomeIcon icon={faAngleRight}/></a>
-            </div>
-            <div style={{minHeight: '200px', margin: '0 auto'}}>
-              <EventCard key={0} event={event} size='l' power={power} />
+        {
+          loadingEvent === 'loading' || loadingEvent === 'idle' ?
+              <div style={{display: 'flex', justifyContent: 'center'}}>
```
Comment (pablojp) [inline src/pages/event.js:223]: remove in line style and move to css files with a class

File: src/pages/event.js [hunk 9]
```diff
@@ -190,77 +218,71 @@ export function Event() {
...
+                <Loader />
+              </div> :
+          errorEvent || Object.keys(event).length === 0 ?
+            <div style={{display: 'flex', justifyContent: 'center', flexDirection: 'column', margin: '0 auto', textAlign: 'center'}}>
```
Comment (pablojp) [inline src/pages/event.js:227]: Can you move this div to an inner component?

Comment (pablojp) [reply src/pages/event.js:227]: and move inline styles

File: src/pages/event.js [hunk 10]
```diff
@@ -190,77 +218,71 @@ export function Event() {
...
+              <h2>{errorEvent || 'Token not found'}</h2>
```
Comment (pablojp) [inline src/pages/event.js:228]: `'Token not found'` is a wording to render in the page? need move to a constant

File: src/pages/event.js [hunk 11]
```diff
@@ -48,10 +60,15 @@ export function Event() {
   const [data, setData] = useState([]);
   const [csv_data, setCsv_data] = useState([]);
   const [ensNames, setEnsNames] = useState([]);
+  const [canDownloadCsv, setCanDownloadCsv] = useState(CSV_STATUS.DownloadingData);
   const width = useWindowWidth();
   const pageCount = useMemo( () => event.tokenCount % 50 !== 0 ? Math.floor(event.tokenCount / 50) + 1 : event.tokenCount, [event])
   const power = calculatePower(csv_data);
 
+  const csvDownloadIsOnLastStep = () => {return canDownloadCsv === CSV_STATUS.DownloadingLastDataChunk}
```
Comment (pablojp) [inline src/pages/event.js:68]:
```suggestion
  const csvDownloadIsOnLastStep = () => canDownloadCsv === CSV_STATUS.DownloadingLastDataChunk
```

File: src/pages/event.js [hunk 12]
```diff
@@ -48,10 +60,15 @@ export function Event() {
...
+  const csvReadyOrAlmostReady = () => {return canDownloadCsv === CSV_STATUS.Ready || canDownloadCsv === CSV_STATUS.ReadyWithoutEns}
```
Comment (pablojp) [inline src/pages/event.js:69]:
```suggestion
  const csvReadyOrAlmostReady = () => canDownloadCsv === CSV_STATUS.Ready || canDownloadCsv === CSV_STATUS.ReadyWithoutEns
```

File: src/pages/event.js [hunk 13]
```diff
@@ -48,10 +60,15 @@ export function Event() {
...
+  const csvOnlyMissingEns = () => {return canDownloadCsv === CSV_STATUS.ReadyWithoutEns}
```
Comment (pablojp) [inline src/pages/event.js:70]:
```suggestion
  const csvOnlyMissingEns = () => canDownloadCsv === CSV_STATUS.ReadyWithoutEns
```

File: src/pages/event.js [hunk 14]
```diff
@@ -190,77 +218,71 @@ export function Event() {
...
+                      filename={`${event.name}.csv`}
+                      target="_blank"
+                      data-tip={`${csvOnlyMissingEns() ? 'Please wait if you want the ens names too' : ''}`}
+                      className={'btn'}
+                      style={csvButtonStyle}
+                      data={csv_data}
+                  >
+                    <CSVLinkText>{`Download CSV${csvOnlyMissingEns() ? ' (without ens)' : ''}`}</CSVLinkText>
+                    <ReactTooltip effect={'solid'} />
+                  </CSVLink> :
```
Comment (pablojp) [inline src/pages/event.js:266]: this double conditional with `:`, in this case, increases the cognitive load of the code and adds a high probability of error by not detecting that it is a double conditional. Be very careful in the future use of these conditionals, they are only advisable in cases of one line of code.
The most advisable thing now is to separate into two conditionals, even if it is the same but denied.

File: src/pages/event.js [hunk 15]
```diff
@@ -190,77 +218,71 @@ export function Event() {
...
+                      data-tip={'Please wait for the POAPs data to be loaded'}
+                      onClick={null}
+                  ><span style={{margin: 0}}>Download CSV</span><ReactTooltip effect={'solid'} /></button>
+                }
+              </div>
+              <div className='table-container'>
+                {
+                  width > 480
+                      ? <CreateTable event={event} loading={loadingEvent !== 'succeeded'} columns={columns} data={data} pageCount={pageCount} />
+                      : <CreateMobileTable event={event} loading={loadingEvent !== 'succeeded'} columns={mobileColumns} data={data} pageCount={pageCount} />
```
Comment (pablojp) [inline src/pages/event.js:279]: If you see lines 278 and 279 they are practically identical. In this case we can use an adapter pattern for a more verbose code
 
```js
{   const isLoading = loadingEvent !== 'succeeded'
    const AdapterCreateTable = (width > 480) ? CreateTable : CreateMobileTable
    <AdapterCreateTable event={event} 
                        loading={isLoading} 
                        columns={mobileColumns} 
                        data={data} 
                        pageCount={pageCount} />
}    
```

File: src/pages/gallery.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (pablojp) [reply src/pages/gallery.js:52]:
```js
    setSearch(undefined);
    const hasMoreThan3Characters = value && value.length > 2;
    
    if (hasMoreThan3Characters) {
      const filteredItems = items.filter((item) => {
        return item.name.toLowerCase().indexOf(value.toLowerCase()) !== -1;
      });
      setSearch(filteredItems);
    }
```

File: src/pages/event.js [hunk 16]
```diff
@@ -65,19 +82,22 @@ export function Event() {
...
 
   useEffect(() => {
-    let ownerIds = tokens.map(t => t.owner.id)
-    if (event && event.tokenCount > GRAPH_LIMIT && tokens && tokens.length > 0) {
-      const totalPages = Math.ceil(event.tokenCount / GRAPH_LIMIT);
-      if (pageIndex + 1 < totalPages) {
-        setPageIndex(pageIndex + 1);
+    // Call next batch of tokens (if there is more), then load the new tokens data
+    const totalPages = Math.ceil(event.tokenCount / GRAPH_LIMIT);
+    if (event && tokens && tokens.length > 0 && pageIndex < totalPages) {
```
Comment (pablojp) [inline src/pages/event.js:96]: Move this conditional to one or more constants with an appropriate name according to your responsibility.

File: src/pages/event.js [hunk 17]
```diff
@@ -163,25 +197,19 @@ export function Event() {
...
-          <meta property="og:url" content={"https://poap.gallery/event/" + eventId}/>
-          <meta property="og:title" content="POAP Gallery - Event"/>
-        </Helmet>
-        <Foliage />
-        <div style={{display: 'flex', justifyContent: 'center'}}>
-          <Loader />
-        </div>
-      </main>
-    )
+  const csvButtonStyle = {
```
Comment (pablojp) [inline src/pages/event.js:200]: move to css class

File: src/pages/event.js [hunk 18]
```diff
@@ -190,77 +218,71 @@ export function Event() {
...
-              <h4 style={{marginBottom: '0'}}><div className='event-title'>EVENT ID</div><div className='event-id'>#{eventId}</div> </h4>
-              <a href={parseInt(eventId)+1} >{'Next  '}<FontAwesomeIcon icon={faAngleRight}/></a>
-            </div>
-            <div style={{minHeight: '200px', margin: '0 auto'}}>
-              <EventCard key={0} event={event} size='l' power={power} />
+        {
+          loadingEvent === 'loading' || loadingEvent === 'idle' ?
+              <div style={{display: 'flex', justifyContent: 'center'}}>
+                <Loader />
+              </div> :
```
Comment (pablojp) [inline src/pages/event.js:225]: the same case as in line 266

File: src/store/mutations.js [hunk 19]
```diff
@@ -32,8 +34,17 @@ export async function getEnsData(ownerIds){
       console.log(e)
```
Comment (pablojp) [inline src/store/mutations.js:34]: remove this console.log

File: src/store/mutations.js [hunk 20]
```diff
@@ -32,8 +32,15 @@ export async function getEnsData(ownerIds){
...
       names = chunk.map(a => null)
     }
-    const validNames = names.map(name => (namehash.normalize(name) === name && name !== '') && name )
-    allnames = _.concat(allnames, validNames);
+    const validNames = names.map(name => {
+      try {
+        return (namehash.normalize(name) === name && name !== '') && name
+      } catch (e) {
+        console.error('Couldn\'t parse ENS name. ', e)
```
[the remark this answers is not in the document]
Comment (sebasfavaron) [reply src/store/mutations.js]: made it a toast

File: src/pages/event.js [hunk 21]
```diff
@@ -198,77 +221,71 @@ export function Event() {
...
+                      className={'btn button-disabled'}
+                      style={csvButtonStyle}
+                      data-tip={'Please wait for the POAPs data to be loaded'}
+                      onClick={null}
+                  ><span style={{margin: 0}}>Download CSV</span><ReactTooltip effect={'solid'} /></button>
+                }
+              </div>
+              <div className='table-container'>
+                {
+                  width > 480
```
Comment (pablojp) [inline src/pages/event.js:280]: here you can use adapter pattern
```js
{ const isLoading = loadingEvent !== 'succeeded';
  const AdapterCreateTable = (width > 480) ? CreateTable : CreateMobileTable;
  <AdapterCreateTable event={event} 
                      loading={isLoading} 
                      columns={mobileColumns} 
                      data={mobileData} 
                      pageCount={pageCount} />
}
```

File: src/pages/event.js (hunk 6 above)
Comment (pablojp) [reply src/pages/event.js:150]: we can use arrow functions to handle these states
something like that

 ```js
 const isLoadingEvent = () => loadingEvent === LOADING_SUCCEEDED;
```

Comment (sebasfavaron) [reply src/pages/event.js]: https://github.com/poap-xyz/poap-gallery/pull/27/commits/7e947ea870f86e9a75318d72ab194ded2da569a8

File: src/components/eventCard.js [hunk 22]
```diff
@@ -51,6 +52,12 @@ function Header({type, event}) {
     </div>
 }
 
+const DiscreetParagraph = styled.p`
```
Comment (pablojp) [inline src/components/eventCard.js:55]: please move this styles to css file, to remove styles component library.

File: package.json (hunk 5 above)
Comment (pablojp) [reply package.json:47]: please remove it. The idea is to keep the current implementation and not add more variations in the handling of styles in the project

File: src/components/eventCard.js (hunk 22 above)
Comment (pablojp) [reply src/components/eventCard.js:55]: you can replace DiscreetParagraph component with a P html native tag, add a css class and move this styles

File: src/pages/gallery.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (pablojp) [reply src/pages/gallery.js:56]:
```suggestion
    setSearch(undefined); 
    const hasMoreThan3Characters = value && value.length > 2; 
    if (hasMoreThan3Characters) {
```

[the remark this answers is not in the document]
Comment (pablojp) [reply src/pages/gallery.js:56]: and remove the else of line 61

File: src/pages/gallery.js [hunk 23]
```diff
@@ -263,15 +266,15 @@ export default function Gallery() {
...
+            {indexFetchStatus === FETCH_INDEX_PAGE_INFO_STATUS.SUCCEEDED && (
               (search?.length === 0) ? <div className='failed-search'>
                 <img src={FailedSearch} alt='Failed search'/>
                 <h3>No results for that search :(</h3>
               </div> :
               <Cards events={(search?.length) ? search : items} length={search?.length || length} />
-            ) : (
-              <Loader/>
             )}
+            {(indexFetchStatus === FETCH_INDEX_PAGE_INFO_STATUS.IDLE || indexFetchStatus === FETCH_INDEX_PAGE_INFO_STATUS.IDLE) && <Loader/>}
```
Comment (pablojp) [inline src/pages/gallery.js:277]: duplicate condition? `indexFetchStatus === FETCH_INDEX_PAGE_INFO_STATUS.IDLE`
```suggestion
            {(indexFetchStatus === FETCH_INDEX_PAGE_INFO_STATUS.IDLE) && <Loader/>}
```

Comment (sebasfavaron) [reply src/pages/gallery.js:277]: buen ojo! ahi va IDLE y LOADING

**Review by pablojp — APPROVED** [review]

**Pull request merged**